File: examples/perl/replier.pl
#
# Copyright 2007-2016 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.
#



# replier.pl
#
# This is a sample application built with the perl
# GMSEC API.  This application demonstrates several
# key components within the GMSEC API including loading
# from an xml file, producing a heartbeat, and request
# reply functionality using callbacks.  This works in sync
# with Requestor.pl
#  


use strict;
use Util;

use libgmsec_perl;

*isa = \&UNIVERSAL::isa;

# Define RequestReplyCallback Callback
{
	package RequestReplyCallback;
	use base 'libgmsec_perl::Callback';

	*isa = \&UNIVERSAL::isa;

	sub new
	{
		# Initialize the base Package
		my $class = shift;
		my $self = $class->SUPER::new();

		# Initialize member data from arguments
		($self->{CONFIGFILE}) = @_;

		return $self;
	}

	sub onMessage
	{
		my ($self, $connection, $message) = @_;

		my $cfgFile = $self->{CONFIGFILE};

		eval
		{
			# We don't care about anything but request messages
			if ($message->getKind() != $libgmsec_perl::Message::REQUEST)
			{
				return;
			}

			# Obtain the DIRECTIVE-KEYWORD field from the message -- It holds the request count
			my $requestCount = $message->getStringField("DIRECTIVE-KEYWORD")->getValue();
			libgmsec_perl::LogInfo("[RequestReplyCallback::onMessage] Received Request (" . $requestCount . ")");

			# Load the directive response template $message definition
			my $replyTemplate = $cfgFile->lookupMessage("DIRECTIVE-RESPONSE");

			# Create the reply message, with the request message's subject
			my $reply = libgmsec_perl::Message->new($message->getSubject(), $libgmsec_perl::Message::REPLY);

			# Copy fields from the reply message template to our actual reply $message
			$replyTemplate->copyFields($reply);

			# Replace the DATA field with the request number we are responding to
			my $field = libgmsec_perl::StringField->new("DATA", "$requestCount");
			$reply->addField($field);

			# Replace the TIME-COMPLETE field with the correct time
			my $theTime = "YYYY-DDD-HH:MM:SS.sss";
			libgmsec_perl::TimeUtil::formatTime(libgmsec_perl::TimeUtil::getCurrentTime(), $theTime);
			$reply->addField("TIME-COMPLETED", $theTime);

			# Send the reply
			$connection->reply($message, $reply);

			libgmsec_perl::LogInfo("[RequestReplyCallback::onMessage] Reply Sent (" . $requestCount . ")");

			# Find and load the config file log message definition
			my $logMessage = $cfgFile->lookupMessage("LOG-REP");

			# Fill the fields within the log message
			$logMessage->addField("EVENT-TIME", $theTime);
			$logMessage->addField("MSG-TEXT", $requestCount);

			# Publish the log message
			$connection->publish($logMessage);
		};
		if (isa($@, 'libgmsec_perl::Exception'))
		{
			libgmsec_perl::LogError($@->what());
		}
		elsif($@)
		{
			libgmsec_perl::LogError($@);
		}
	}
} # end RequestReplyCallback definition


{
	my $self = bless({});

	# If no arguments were provided, print usage and exit with an erroneous code value
	if (scalar(@ARGV) < 1)
	{
		$self->printUsage();
		exit -1;
	}

	# Initialize logging in the GMSEC API in order to display output
	$self->init();
	
	eval
	{
		$self->run($ARGV[0]);
	};
	if ($@)
	{
		print "caught: $@";
	}
}


sub printUsage
{
	print "\nusage: perl replier.pl <filename>\n";
}


sub init
{
	my $config = libgmsec_perl::Config->new();
	$config->addValue("LOGLEVEL", "INFO");
}


sub run
{
	my ($self, $cfgFileName) = @_;

	eval
	{
		# Load the config file
		my $cfgFile = libgmsec_perl::ConfigFile->new();
		$cfgFile->load($cfgFileName);

		# Look up configuration information
		my $progConfig    = $cfgFile->lookupConfig("program-config");
		my $connConfig    = $cfgFile->lookupConfig("connection-config");
		my $heartbeatMsg  = $cfgFile->lookupMessage("C2CX-HEARTBEAT-REP");

		# Get program config values
		my $updateRate    = Util::getInteger($progConfig, "update-rate", 1);
		my $loopCountdown = Util::getInteger($progConfig, "loop-time", 30);
		my $hbPubRate     = $heartbeatMsg->getI16Field("PUB-RATE")->getValue();

		# Create a $connection object off of the configuration
		my $connection = libgmsec_perl::Connection::create($connConfig);

		# Connect to the GMSEC bus
		$connection->connect();

		# Output version information
		libgmsec_perl::LogInfo("API version = " . libgmsec_perl::Connection::getAPIVersion());
		libgmsec_perl::LogInfo("Middleware version = " . $connection->getLibraryVersion());

		# Create subscriptions from subscription templates in the config file using callback
		my $cb = RequestReplyCallback->new($cfgFile);
		my $topic = $cfgFile->lookupSubscription("DIRECTIVE-REQUEST");

		$self->{SUBINFO} = $connection->subscribe($topic, $cb);

		# Output some general program information
		libgmsec_perl::LogInfo("Publishing for " . $loopCountdown . " seconds->");
		libgmsec_perl::LogInfo("Publishing Heartbeat Messages every " . $hbPubRate . " seconds->");

		# Call to create the thread that will automatically retrieve $messages off the message queue
		$connection->startAutoDispatch();

		# Display start time
		my $theTime = "YYYY-DDD-HH:MM:SS.sss";
		libgmsec_perl::TimeUtil::formatTime(libgmsec_perl::TimeUtil::getCurrentTime(), $theTime);
		libgmsec_perl::LogInfo("Start Time: " . $theTime);

		# Publishing loop
		my $heartbeatCount     = 0;
		my $heartbeatCountdown = 0;

		for (; $loopCountdown > 0; --$loopCountdown)
		{
			# When the countdown reaches 0, we publish a heartbeat $message and reset the counter
			if ($heartbeatCountdown < 1)
			{
				++$heartbeatCount;

				$heartbeatMsg->addField("COUNTER", $heartbeatCount);

				# Publish the heartbeat $message
				$connection->publish($heartbeatMsg);

				# Output a heartbeat marker and reset the counter
				libgmsec_perl::LogInfo("Published Heartbeat");
				$heartbeatCountdown = $hbPubRate;
			}

			# Decrement the counters
			$heartbeatCountdown -= $updateRate;

			# Sleep
			Util::millisleep($updateRate * 1000);
		}

		# End time
		libgmsec_perl::TimeUtil::formatTime(libgmsec_perl::TimeUtil::getCurrentTime(), $theTime);
		libgmsec_perl::LogInfo("End Time: " . $theTime);

		$connection->stopAutoDispatch();
	};
	if (isa($@, 'libgmsec_perl::Exception'))
	{
		libgmsec_perl::LogError($@->what());
	}
	elsif($@)
	{
		libgmsec_perl::LogError($@);
	}

	$self->cleanup();
}


sub cleanup
{
	my ($self) = @_;

	my $connection = $self->{CONNECTION};

	if ($connection)
	{
		my $subInfo = $self->{SUBINFO};

		eval
		{
			libgmsec_perl::LogInfo("Unsubscribing from " . $subInfo->getSubject());
			$connection->unsubscribe($subInfo);
		};
		if (isa($@, 'libgmsec_perl::Exception'))
		{
			libgmsec_perl::LogError($@->what());
		}
		elsif($@)
		{
			libgmsec_perl::LogError($@);
		}

		libgmsec_perl::Connection::destroy($connection);
	}

	libgmsec_perl::Connection::shutdownAllMiddlewares();
}


1;

__END__

File: examples/java/Replier.java
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/** 
 * @file Replier.java
 *
 *  A simple examples of various GMSEC API functionality.
 *
 * Contained within this example:
 *
 * - Loading an XML configuration file
 * - Populating a Config object using the XML file
 * - Establish a connection
 * - Using the XML file to store and set various other program configurations
 * - Subscribing to specific subjects to receieve messages through a callback
 * - Creating a heartbeat message from an XML definition
 * - Starting the automatic message dispatching thread
 * - Publishing the heartbeat message using the Publish call
 * - Programatically altering a field in the heartbeat message
 * Creating a response message from an XML definition upon receiving a request
 * Programatically altering a field in the response message
 * Using the Reply call to send a reply message
 * - Creating a log message from an XML definition
 * - Programatically altering a field in the log message
 * - Publishing the log message using the Publish call
 * Cleanup
 * - Message memory cleanup
 * - Connection cleanup disconnectiong from the network
 *
 * Any item with a * next to it is specific to this file in the overall example.
 */


import gov.nasa.gsfc.gmsec.api.*;
import gov.nasa.gsfc.gmsec.api.field.I16Field;
import gov.nasa.gsfc.gmsec.api.util.Log;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;


class ReplyInnerExample implements Example
{
	ConfigFile configFile;
	Connection connection;
	Message    message;


	ReplyInnerExample(ConfigFile configFile, Connection connection, Message message)
	{
		this.configFile = configFile;
		this.connection = connection;
		this.message    = message;
	}


	public boolean run()
	{
		boolean result = true;
		String  now    = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date());

		try
		{
			String dirKey = message.getStringField("DIRECTIVE-KEYWORD").getValue();

			Log.info("Received Request: " + dirKey);

			// Lookup and load the config file directive response message definition
			Message tmpMsg = configFile.lookupMessage("DIRECTIVE-RESPONSE");

			// Replace the DATA and TIME-COMPLETE fields within the temporary response message
			tmpMsg.addField("DATA", dirKey);
			tmpMsg.addField("TIME-COMPLETED", now);

			// Create the actual reply message
			Message replyMessage = new Message(message.getSubject(), Message.MessageKind.REPLY);

			// Copy the fields from the temporary response message to our reply message
			tmpMsg.copyFields(replyMessage);

			// Send the reply
			connection.reply(message, replyMessage);

			Log.info("Reply sent.");

			// Lookup and load the config file log message definition
			Message logMessage = configFile.lookupMessage("LOG-REP");

			// Fill the log message time and text
			logMessage.addField("EVENT-TIME", now);
			logMessage.addField("MSG-TEXT", dirKey);

			// Publish the log message
			connection.publish(logMessage);
		}
		catch (GMSEC_Exception e)
		{
			Log.error("GMSEC_Exception: " + e.toString());
			result = false;
		}

		return result;
	}


	public boolean cleanup()
	{
		return true;
	}
}


// Callback Class for Request-Reply
class RequestReplyCallback extends Callback
{
	private ConfigFile configFile;


	public RequestReplyCallback(ConfigFile cfgFile)
	{
		configFile = cfgFile;
	}


	public void onMessage(Connection conn, Message msg)
	{
		// We don't care about anything but request messages
		if (msg.getKind() != Message.MessageKind.REQUEST)
		{
			return;
		}

		new ReplyInnerExample(configFile, conn, msg).run();
	}
}


public class Replier implements Example
{
	Config     config;
	ConfigFile configFile;
	Connection conn;
	ArrayList<SubscriptionInfo> info = new ArrayList<SubscriptionInfo>();


	Replier(Config config, ConfigFile configFile)
	{
		this.config     = config; 
		this.configFile = configFile;
		this.conn       = null;

		Util.initialize(config);
	}


	public boolean run()
	{
		boolean result = true;

		try
		{
			// Establish a connection object off of the configuration
			conn = Util.openConnection(config); 

			// output GMSEC API and Middleware info, just for display purposes
			Log.info(Connection.getAPIVersion());
			Log.info(conn.getLibraryVersion());

			// Look up additional program configutation information
			Config progConfig = configFile.lookupConfig("program-config");

			// Get program config values
			short updateRate    = Short.parseShort(Util.get(progConfig, "update-rate"));
			short loopCountdown = Short.parseShort(Util.get(progConfig, "loop-time"));

			// Create subscriptions from subscription templates in the config file
			String subject = configFile.lookupSubscription("DIRECTIVE-REQUEST");

			// Subscribe
			info.add(conn.subscribe(subject, new RequestReplyCallback(configFile)));

			// Find and load the config file heartbeat message definition
			Message heartbeatMessage = configFile.lookupMessage("C2CX-HEARTBEAT-REP");

			// Obtain the publish rate field, by name, from the heartbeat message
			short heartbeatHolder = 30;

			try {
				heartbeatHolder = heartbeatMessage.getI16Field("PUB-RATE").getValue();
			}
			catch (GMSEC_Exception e) {
				Log.warning("PUB-RATE not found in C2CX-HEARTBEAT-REP message; using default rate of 30");
			}

			// Main Publishing Loop

			// Output some general program information
			Log.info("Publishing for " + loopCountdown + " seconds.");
			Log.info("Publishing Heartbeat Messages every " + heartbeatHolder + " seconds.");

			// Call to create the thread that will automatically retrieve messages 
			// off the message queue
			conn.startAutoDispatch();

			// Start time
			String startTime = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date());
			Log.info("Start Time: " + startTime);

			// Publishing loop
			short heartbeatCount = 0;
			short heartbeatCountdown = 0;

			for (;loopCountdown > 0; --loopCountdown)
			{
				// When the countdown reaches 0, we publish a heartbeat message and
				//  reset the counter
				if (heartbeatCountdown < 1)
				{
					++heartbeatCount;

					// Update the message counter
					heartbeatMessage.addField(new I16Field("COUNTER", heartbeatCount));

					// Publish the heartbeat message
					conn.publish(heartbeatMessage);

					// Ouput a heartbeat marker and reset the counter
					Log.info("Published heartbeat...");

					heartbeatCountdown = heartbeatHolder;
				}

				// Decrement the counters
				heartbeatCountdown = (short)(heartbeatCountdown - updateRate);

				// Sleep for 1 second
				Util.millisleep(1000);
			}

 			// End time
			 String endTime = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date());
			 Log.info("End Time: " + endTime);
  
			// Stop auto dispatcher
			conn.stopAutoDispatch();
		}
		catch (GMSEC_Exception e)
		{
			Log.error("GMSEC_Exception: " + e.toString());
			result = false;
		}
		finally
		{
			try
			{
				cleanup();
			}
			catch (GMSEC_Exception e)
			{
				Log.error("GMSEC_Exception: " + e.toString());
				result = false;
			}
		}
  
		return result;
 	}


	public boolean cleanup() throws GMSEC_Exception
	{
		if (conn != null)
		{
			for(int i = info.size()-1; i >= 0; i-- )
			{
				Log.info("Unsubscribing from " + info.get(i).getSubject());
				conn.unsubscribe(info.get(i));
				info.remove(i);
			}
			Util.closeConnection(conn);
		}

		return true;
	}


	public static void main(String args[])
	{
		// Simple check for proper usage
		if (args.length != 1)
		{
			System.err.println("\nUsage:  java Replier <XML config filename>\n\n");
			System.exit(-1);
		}

		try
		{
			// Create the config file and associate the config filename
			ConfigFile configFile = new ConfigFile();

			configFile.load(args[0]);

			// Look up the config file value that contains the configuration for the middleware
			Config config = configFile.lookupConfig("connection-config");

			new Replier(config, configFile).run();
		}
		catch (GMSEC_Exception e)
		{
			System.err.println("GMSEC_Exception: " + e.toString());
			System.exit(-1);
		}
	}
}
